design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,478 @@
+specification already looks and feels a lot like the Kubernetes API. When it is
+implemented using the _actual_ Kubernetes API any divergences from said API’s
+conventions are an order of magnitude more likely to violate the [principle of
+least astonishment] for users than for users of other runtimes.
+
+Any proposal outlined by this document should ensure:
+
+* The separation of concerns that is the core focus of the OAM specification.
+* That the OAM specification “look and feel” idiomatic to the Kubernetes API
+  model.
+* That the OAM specification remains useful and compelling to other runtimes.
+
+Note that this document frequently uses Crossplane as an example of a tentative
+OAM runtime that would benefit from a Kubernetes friendly OAM specification, but
+defining specifically how Crossplane will support OAM is out of scope.
+
+## Proposal
+
+This document proposes:
+
+1. An ApplicationConfiguration _references_ distinct Component resources,
+   grouping them into one logical application and providing inputs to their
+   parameters.
+1. Each Component embed an OAM “workload” under the `.spec.workload` key.
+   Available workloads are defined by WorkloadDefinitions. In Kubernetes
+   runtimes any existing custom resource with a corresponding WorkloadDefinition
+   can be used in a Component.
+1. The set of available Traits and their schemas be defined via
+   TraitDefinitions. Traits are applied to a component and configured by
+   specifying their configuration inline in the ApplicationConfig as an
+   "embedded resource".
+1. The set of available Scopes and their schemas be defined via
+   ScopeDefinitions. Scopes are referenced by an ApplicationConfiguration.
+
+### Components and Workloads
+
+Each Component is of a distinct workload _kind_. This is functionally equivalent
+to the existing concept of workload _types_, but reuses the _kind_ term already
+used within Kubernetes and OAM. The schema of a workload kind - including core
+workload kinds like Server - are defined via a WorkloadDefinition. For example:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: WorkloadDefinition
+metadata:
+  name: server.core.oam.dev
+spec:
+  definitionRef:
+    name: server.core.oam.dev
+```
+
+The above WorkloadDefinition references a Kubernetes CustomResourceDefinition
+named server.core.oam.dev, designating custom resources defined by said CRD to
+be valid OAM workloads. OAM runtimes that are not powered by Kubernetes could
+opt to either implement the CRD concept, or introduce a new field to the
+WorkloadDefinition spec that defines a kind of workload and its schema.
+Application operators can discover available OAM workload kinds by running
+`kubectl get workloaddefinitions`. Workloads would most frequently be defined by
+OAM runtimes or infrastructure operators.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: Component
+metadata:
+  name: example-server
+spec:
+  workload:
```

Comment:
@vturecek @hongchaodeng @ryanzhang-oss @resouer - This morning we spoke in the community meeting about how application operators would specify traits in the OAM ApplicationConfiguration. I believe we decided that the inline traits would need to be somehow relatable back to their trait definitions (e.g. by including some kind of type metadata), but would not explicitly need to include `kind` and `apiVersion` metadata.

How does this relate to workload kinds, especially core ones such as `Server`? We're proposing here that the `Component` embed a core workload type, which looks like a Kubernetes CR. The workload's kind is defined by the `apiVersion` and `kind` fields of the embedded CR. My thinking was that these core workload kinds would look like a Kubernetes CR (with `apiVersion` and `kind) even if the OAM runtime were not Kubernetes based. This allows portability of `Component` instances across runtimes - which I think was @abchow's concern this morning.

Reply:
The indexing mechanism from `workload -> def'n` and `trait -> def'n` is same IMO.

Reply:
@resouer It's not clear to me what you mean.

My point here is that we did not seem comfortable with explicitly requiring traits to include an `apiVersion` and `kind` in their schema, yet we seem to be doing that implicitly for workloads. A workload on Kubernetes is an embedded CR, and CRs must have `apiVersion` and `kind`. We want to have a set of core workloads (like the `kind: Server` example here) that are portable across runtimes. It therefore follows that at least our core workloads _must_ have `apiVersion` and `kind` in their schema regardless of what platform they're on. Why not take this one step further and require that all traits, scopes, and workloads contain `apiVersion` and `kind` regardless of runtime?

Reply:
It's not the intention to make them different. So what I mean is as long as we agree on one approach, it's reasonable applies to both workload and trait (and other xxxDefinition enforced concepts)